File: task-e37916/solution/xv6_fsck.py
#!/usr/bin/env python3

"""
xv6 Filesystem Consistency Checker (fsck)

Parses a raw xv6 filesystem image and detects inconsistencies in the
superblock, block bitmap, inode block references, directory entries,
and nlink counts.

Usage: xv6-fsck <image_path>
Output: JSON report to stdout
"""

import struct
import sys
import json

# ── xv6 filesystem constants ──
BSIZE = 1024
NDIRECT = 12
NINDIRECT = BSIZE // 4   # 256
IPB = BSIZE // 64        # 16  (inodes per block)
BPB = BSIZE * 8          # 8192 (bitmap bits per block)
T_DIR = 1
T_FILE = 2
T_DEVICE = 3
FSMAGIC = 0x10203040
DIRSIZ = 14
ROOTINO = 1


def read_superblock(data):
    """Parse the superblock from block 1."""
    fields = struct.unpack_from('<IIIIIIII', data, BSIZE)
    return {
        'magic': fields[0],
        'size': fields[1],
        'nblocks': fields[2],
        'ninodes': fields[3],
        'nlog': fields[4],
        'logstart': fields[5],
        'inodestart': fields[6],
        'bmapstart': fields[7],
    }


def compute_datastart(sb):
    """Compute the first data block number from superblock fields."""
    niblocks = (sb['ninodes'] + IPB - 1) // IPB
    nbmapblocks = (sb['size'] + BPB - 1) // BPB
    return sb['bmapstart'] + nbmapblocks


def read_inode(data, inum, sb):
    """Read an on-disk inode (dinode) by inode number."""
    block = inum // IPB + sb['inodestart']
    off = (inum % IPB) * 64
    pos = block * BSIZE + off
    typ, major, minor, nlink, size = struct.unpack_from('<hhhhI', data, pos)
    addrs = list(struct.unpack_from('<13I', data, pos + 12))
    return {
        'type': typ,
        'major': major,
        'minor': minor,
        'nlink': nlink,
        'size': size,
        'addrs': addrs,
    }


def get_bitmap_bit(data, block, sb):
    """Return 1 if bitmap says block is allocated, 0 otherwise."""
    bmap_block_idx = block // BPB
    bit = block % BPB
    byte_offset = (sb['bmapstart'] + bmap_block_idx) * BSIZE + bit // 8
    if byte_offset >= len(data):
        return 0
    return (data[byte_offset] >> (bit % 8)) & 1


def collect_inode_blocks(data, inode, sb, datastart):
    """
    Collect all block numbers referenced by an inode.
    Returns list of (block_number, is_valid) tuples.
    A block is valid if it's in range [datastart, sb['size']).
    Block number 0 means "not allocated" and is skipped.
    """
    blocks = []

    # Direct blocks
    for i in range(NDIRECT):
        b = inode['addrs'][i]
        if b != 0:
            blocks.append(b)

    # Indirect block
    indirect_ptr = inode['addrs'][NDIRECT]
    if indirect_ptr != 0:
        blocks.append(indirect_ptr)  # The indirect block itself is referenced

        # Read entries from the indirect block if it's accessible
        if indirect_ptr < sb['size']:
            indirect_pos = indirect_ptr * BSIZE
            if indirect_pos + BSIZE <= len(data):
                for j in range(NINDIRECT):
                    entry = struct.unpack_from('<I', data, indirect_pos + j * 4)[0]
                    if entry != 0:
                        blocks.append(entry)

    return blocks


def read_directory_entries(data, inode, sb):
    """
    Read all directory entries from a directory inode.
    Returns list of {'inum': int, 'name': str} for non-free entries.
    """
    entries = []
    if inode['type'] != T_DIR:
        return entries

    size = inode['size']
    offset = 0

    while offset < size:
        # Determine which block this offset falls in
        block_idx = offset // BSIZE
        offset_in_block = offset % BSIZE

        if block_idx < NDIRECT:
            block_addr = inode['addrs'][block_idx]
        elif inode['addrs'][NDIRECT] != 0:
            indirect_ptr = inode['addrs'][NDIRECT]
            if indirect_ptr >= sb['size'] or indirect_ptr * BSIZE + BSIZE > len(data):
                offset += 16
                continue
            indirect_pos = indirect_ptr * BSIZE
            inner_idx = block_idx - NDIRECT
            if inner_idx >= NINDIRECT:
                break
            block_addr = struct.unpack_from('<I', data, indirect_pos + inner_idx * 4)[0]
        else:
            offset += 16
            continue

        if block_addr == 0 or block_addr >= sb['size'] or block_addr * BSIZE + BSIZE > len(data):
            offset += 16
            continue

        pos = block_addr * BSIZE + offset_in_block
        if pos + 16 > len(data):
            break

        inum = struct.unpack_from('<H', data, pos)[0]
        name_raw = data[pos + 2:pos + 16]

        if inum != 0:
            # Decode name: find NUL or take all DIRSIZ bytes
            null_pos = name_raw.find(0)
            if null_pos >= 0:
                name = name_raw[:null_pos].decode('ascii', errors='replace')
            else:
                name = name_raw.decode('ascii', errors='replace')
            entries.append({'inum': inum, 'name': name})

        offset += 16

    return entries


def fsck(image_path):
    """Run all consistency checks on the given xv6 filesystem image."""
    with open(image_path, 'rb') as f:
        data = bytearray(f.read())

    sb = read_superblock(data)
    errors = []

    # Validate magic number
    if sb['magic'] != FSMAGIC:
        return {'errors': [{'type': 'BAD_MAGIC'}]}

    datastart = compute_datastart(sb)

    # ── Pass 1: Read all inodes and collect block references ──
    all_inodes = {}
    block_refs = {}  # block -> list of inodes referencing it

    for inum in range(sb['ninodes']):
        inode = read_inode(data, inum, sb)
        all_inodes[inum] = inode

        if inode['type'] == 0:
            continue

        blocks = collect_inode_blocks(data, inode, sb, datastart)
        for b in blocks:
            if b not in block_refs:
                block_refs[b] = []
            block_refs[b].append(inum)

    # ── Check 1: Invalid block references ──
    for block, inodes in block_refs.items():
        if block >= sb['size'] or block < datastart:
            for inum in inodes:
                errors.append({
                    'type': 'INVALID_BLOCK_REF',
                    'inode': inum,
                    'block': block,
                })

    # ── Check 2: Duplicate block references ──
    for block, inodes in block_refs.items():
        if len(inodes) > 1 and datastart <= block < sb['size']:
            errors.append({
                'type': 'DUPLICATE_BLOCK_REF',
                'block': block,
                'inodes': sorted(inodes),
            })

    # ── Check 3: Bitmap consistency (data blocks only) ──
    for b in range(datastart, sb['size']):
        bitmap_set = get_bitmap_bit(data, b, sb)
        referenced = b in block_refs

        if referenced and not bitmap_set:
            # Only report if the block is in valid data range
            # (invalid refs are already caught above)
            ref_inodes = block_refs[b]
            # Only include inodes whose reference to this block is valid (data range)
            if datastart <= b < sb['size']:
                errors.append({
                    'type': 'BITMAP_MARKED_FREE',
                    'block': b,
                    'inodes': sorted(ref_inodes),
                })
        elif bitmap_set and not referenced:
            errors.append({
                'type': 'BITMAP_MARKED_USED',
                'block': b,
            })

    # ── Pass 2: Walk directories to count nlink references ──
    dir_refs = {}  # inum -> count of directory entries pointing to it (excl ".")

    for inum in range(sb['ninodes']):
        inode = all_inodes[inum]
        if inode['type'] != T_DIR:
            continue

        entries = read_directory_entries(data, inode, sb)
        for entry in entries:
            if entry['name'] == '.':
                continue  # "." does not contribute to nlink
            target = entry['inum']
            if target not in dir_refs:
                dir_refs[target] = 0
            dir_refs[target] += 1

    # ── Check 4: Orphan inodes ──
    for inum in range(1, sb['ninodes']):  # Skip inode 0 (always unused)
        inode = all_inodes[inum]
        if inode['type'] == 0:
            continue

        ref_count = dir_refs.get(inum, 0)
        if ref_count == 0:
            errors.append({
                'type': 'ORPHAN_INODE',
                'inode': inum,
            })

    # ── Check 5: nlink mismatches ──
    for inum in range(1, sb['ninodes']):
        inode = all_inodes[inum]
        if inode['type'] == 0:
            continue

        expected = dir_refs.get(inum, 0)
        actual = inode['nlink']

        if expected != actual:
            errors.append({
                'type': 'INODE_NLINK_MISMATCH',
                'inode': inum,
                'expected': expected,
                'actual': actual,
            })

    return {'errors': errors}


def main():
    if len(sys.argv) < 2:
        print("Usage: xv6-fsck <image_path>", file=sys.stderr)
        sys.exit(1)

    image_path = sys.argv[1]
    report = fsck(image_path)
    print(json.dumps(report, indent=2))


if __name__ == '__main__':
    main()

File: task-e37916/solution/xv6_fsck.c
/*
 * xv6 Filesystem Consistency Checker
 * Parses a raw xv6 filesystem image and detects structural inconsistencies.
 * Usage: xv6-fsck <image_path>
 * Output: JSON report to stdout
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

/* xv6 filesystem constants */
#define BSIZE 1024
#define NDIRECT 12
#define NINDIRECT 256
#define IPB 16
#define BPB 8192
#define FSMAGIC 0x10203040
#define T_DIR 1
#define DIRSIZ 14

static uint8_t *disk;
static size_t disk_len;

static uint32_t sb_magic, sb_size, sb_nblocks, sb_ninodes;
static uint32_t sb_nlog, sb_logstart, sb_inodestart, sb_bmapstart;
static uint32_t datastart;

/* Block reference tracking: (block, inum) pairs */
struct bref {
    uint32_t block;
    uint32_t inum;
};

static struct bref *refs;
static int nrefs, rcap;

/* Directory reference counts per inode */
static int *dref;

/* JSON output state */
static int first_err;

static void addref(uint32_t block, uint32_t inum)
{
    if (nrefs >= rcap) {
        rcap = rcap ? rcap * 2 : 4096;
        refs = realloc(refs, (size_t)rcap * sizeof(*refs));
        if (!refs) { perror("realloc"); exit(1); }
    }
    refs[nrefs].block = block;
    refs[nrefs].inum = inum;
    nrefs++;
}

static void read_superblock(void)
{
    if (disk_len < 2 * BSIZE) {
        fprintf(stderr, "Image too small for superblock\n");
        exit(1);
    }
    memcpy(&sb_magic, disk + BSIZE, 4);
    memcpy(&sb_size, disk + BSIZE + 4, 4);
    memcpy(&sb_nblocks, disk + BSIZE + 8, 4);
    memcpy(&sb_ninodes, disk + BSIZE + 12, 4);
    memcpy(&sb_nlog, disk + BSIZE + 16, 4);
    memcpy(&sb_logstart, disk + BSIZE + 20, 4);
    memcpy(&sb_inodestart, disk + BSIZE + 24, 4);
    memcpy(&sb_bmapstart, disk + BSIZE + 28, 4);
}

static int get_inode_type(uint32_t inum)
{
    size_t pos = ((size_t)(inum / IPB + sb_inodestart)) * BSIZE + (inum % IPB) * 64;
    if (pos + 64 > disk_len) return 0;
    int16_t type;
    memcpy(&type, disk + pos, 2);
    return type;
}

static int16_t get_inode_nlink(uint32_t inum)
{
    size_t pos = ((size_t)(inum / IPB + sb_inodestart)) * BSIZE + (inum % IPB) * 64;
    if (pos + 64 > disk_len) return 0;
    int16_t nlink;
    memcpy(&nlink, disk + pos + 6, 2);
    return nlink;
}

static void get_inode_addrs(uint32_t inum, uint32_t addrs[NDIRECT + 1])
{
    size_t pos = ((size_t)(inum / IPB + sb_inodestart)) * BSIZE + (inum % IPB) * 64;
    memcpy(addrs, disk + pos + 12, (NDIRECT + 1) * 4);
}

static uint32_t get_inode_size(uint32_t inum)
{
    size_t pos = ((size_t)(inum / IPB + sb_inodestart)) * BSIZE + (inum % IPB) * 64;
    uint32_t sz;
    memcpy(&sz, disk + pos + 8, 4);
    return sz;
}

static int get_bitmap(uint32_t b)
{
    size_t off = ((size_t)(sb_bmapstart + b / BPB)) * BSIZE + (b % BPB) / 8;
    if (off >= disk_len) return 0;
    return (disk[off] >> ((b % BPB) % 8)) & 1;
}

static void collect_blocks(uint32_t inum)
{
    if (get_inode_type(inum) == 0) return;

    uint32_t addrs[NDIRECT + 1];
    get_inode_addrs(inum, addrs);

    /* Direct blocks */
    for (int i = 0; i < NDIRECT; i++) {
        if (addrs[i] != 0)
            addref(addrs[i], inum);
    }

    /* Indirect block */
    uint32_t ind = addrs[NDIRECT];
    if (ind != 0) {
        addref(ind, inum);
        if (ind < sb_size && (size_t)ind * BSIZE + BSIZE <= disk_len) {
            uint32_t ents[NINDIRECT];
            memcpy(ents, disk + (size_t)ind * BSIZE, sizeof(ents));
            for (int j = 0; j < NINDIRECT; j++) {
                if (ents[j] != 0)
                    addref(ents[j], inum);
            }
        }
    }
}

static void walk_directory(uint32_t inum)
{
    if (get_inode_type(inum) != T_DIR) return;

    uint32_t sz = get_inode_size(inum);
    uint32_t addrs[NDIRECT + 1];
    get_inode_addrs(inum, addrs);

    uint32_t off = 0;
    while (off < sz) {
        uint32_t bi = off / BSIZE;
        uint32_t boff = off % BSIZE;
        uint32_t ba;

        if (bi < NDIRECT) {
            ba = addrs[bi];
        } else {
            uint32_t ind = addrs[NDIRECT];
            if (ind == 0 || ind >= sb_size) { off += 16; continue; }
            uint32_t inner = bi - NDIRECT;
            if (inner >= NINDIRECT) break;
            memcpy(&ba, disk + (size_t)ind * BSIZE + inner * 4, 4);
        }

        if (ba == 0 || ba >= sb_size) { off += 16; continue; }
        size_t pos = (size_t)ba * BSIZE + boff;
        if (pos + 16 > disk_len) break;

        uint16_t dinum;
        memcpy(&dinum, disk + pos, 2);

        if (dinum != 0 && dinum < sb_ninodes) {
            char nm[DIRSIZ + 1];
            memcpy(nm, disk + pos + 2, DIRSIZ);
            nm[DIRSIZ] = '\0';
            if (strcmp(nm, ".") != 0) {
                dref[dinum]++;
            }
        }
        off += 16;
    }
}

static int cmpref(const void *a, const void *b)
{
    const struct bref *ra = (const struct bref *)a;
    const struct bref *rb = (const struct bref *)b;
    if (ra->block != rb->block)
        return ra->block < rb->block ? -1 : 1;
    if (ra->inum != rb->inum)
        return ra->inum < rb->inum ? -1 : 1;
    return 0;
}

static void emit_sep(void)
{
    if (!first_err) printf(",");
    first_err = 0;
}

int main(int argc, char **argv)
{
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <image>\n", argv[0]);
        return 1;
    }

    /* Read image into memory */
    FILE *f = fopen(argv[1], "rb");
    if (!f) { perror(argv[1]); return 1; }
    fseek(f, 0, SEEK_END);
    disk_len = (size_t)ftell(f);
    rewind(f);
    disk = malloc(disk_len);
    if (!disk) { perror("malloc"); return 1; }
    if (fread(disk, 1, disk_len, f) != disk_len) { perror("fread"); return 1; }
    fclose(f);

    /* Parse superblock */
    read_superblock();
    if (sb_magic != FSMAGIC) {
        printf("{\"errors\":[{\"type\":\"BAD_MAGIC\"}]}\n");
        free(disk);
        return 0;
    }

    uint32_t nbmap = (sb_size + BPB - 1) / BPB;
    datastart = sb_bmapstart + nbmap;

    dref = calloc(sb_ninodes, sizeof(int));
    if (!dref) { perror("calloc"); return 1; }

    /* Pass 1: collect all block references from all inodes */
    for (uint32_t i = 0; i < sb_ninodes; i++)
        collect_blocks(i);

    /* Sort references by (block, inum) for group processing */
    if (nrefs > 0)
        qsort(refs, (size_t)nrefs, sizeof(*refs), cmpref);

    /* Bitmap for tracking which valid data blocks are referenced */
    size_t rmap_bytes = (sb_size + 7) / 8;
    uint8_t *rmap = calloc(rmap_bytes, 1);
    if (!rmap) { perror("calloc"); return 1; }

    first_err = 1;
    printf("{\"errors\":[");

    /* Process sorted block refs: detect invalid, duplicate, bitmap-free */
    int r = 0;
    while (r < nrefs) {
        uint32_t b = refs[r].block;

        /* Collect unique inodes referencing this block */
        uint32_t uniq[512];
        int nu = 0;
        while (r < nrefs && refs[r].block == b) {
            if (nu == 0 || uniq[nu - 1] != refs[r].inum) {
                if (nu < 512)
                    uniq[nu++] = refs[r].inum;
            }
            r++;
        }

        if (b >= sb_size || b < datastart) {
            /* Invalid block reference */
            for (int j = 0; j < nu; j++) {
                emit_sep();
                printf("{\"type\":\"INVALID_BLOCK_REF\",\"inode\":%u,\"block\":%u}",
                       uniq[j], b);
            }
        } else {
            /* Mark block as referenced */
            rmap[b / 8] |= (uint8_t)(1 << (b % 8));

            /* Duplicate block reference */
            if (nu > 1) {
                emit_sep();
                printf("{\"type\":\"DUPLICATE_BLOCK_REF\",\"block\":%u,\"inodes\":[", b);
                for (int j = 0; j < nu; j++) {
                    if (j > 0) printf(",");
                    printf("%u", uniq[j]);
                }
                printf("]}");
            }

            /* Bitmap marked free but block is referenced */
            if (!get_bitmap(b)) {
                emit_sep();
                printf("{\"type\":\"BITMAP_MARKED_FREE\",\"block\":%u,\"inodes\":[", b);
                for (int j = 0; j < nu; j++) {
                    if (j > 0) printf(",");
                    printf("%u", uniq[j]);
                }
                printf("]}");
            }
        }
    }

    /* Bitmap marked used: data blocks with bitmap set but not referenced */
    for (uint32_t b = datastart; b < sb_size; b++) {
        int bm = get_bitmap(b);
        int ref = (rmap[b / 8] >> (b % 8)) & 1;
        if (bm && !ref) {
            emit_sep();
            printf("{\"type\":\"BITMAP_MARKED_USED\",\"block\":%u}", b);
        }
    }

    /* Pass 2: walk all directories to count nlink references */
    for (uint32_t i = 0; i < sb_ninodes; i++)
        walk_directory(i);

    /* Orphan inodes: allocated but no directory entry pointing to them */
    for (uint32_t i = 1; i < sb_ninodes; i++) {
        if (get_inode_type(i) != 0 && dref[i] == 0) {
            emit_sep();
            printf("{\"type\":\"ORPHAN_INODE\",\"inode\":%u}", i);
        }
    }

    /* nlink mismatches */
    for (uint32_t i = 1; i < sb_ninodes; i++) {
        if (get_inode_type(i) == 0) continue;
        int16_t actual_nlink = get_inode_nlink(i);
        if (dref[i] != (int)actual_nlink) {
            emit_sep();
            printf("{\"type\":\"INODE_NLINK_MISMATCH\",\"inode\":%u,\"expected\":%d,\"actual\":%d}",
                   i, dref[i], (int)actual_nlink);
        }
    }

    printf("]}\n");

    free(disk);
    free(dref);
    free(refs);
    free(rmap);
    return 0;
}
